primitives/src/crh/poseidon/sbox.rs

```diff
@@ -0,0 +1,256 @@
+use algebra::{Field, PrimeField, MulShort};
+use crate::{PoseidonParameters, SBox, BatchSBox};
+use std::marker::PhantomData;
+
+pub trait PoseidonInverseParameters: PoseidonParameters {
+    const MDS_CST_SHORT: &'static [Self::Fr];  // The MDS matrix for fast matrix multiplication
+}
+
+pub trait PoseidonSBox<P: PoseidonParameters>: SBox<Field = P::Fr, Parameters = P> {
+
+    // Function that does the scalar multiplication
+    fn scalar_mul(res: &mut <Self as SBox>::Field, state: &mut [<Self as SBox>::Field], start_idx_cst: usize);
+
+    // Function that does the mix matrix
+    fn matrix_mix(state: &mut Vec<<Self as SBox>::Field>)
```

Comment:
As mentioned above: It is unusual to consider the matrix_mix as part of the S-Box layer.

Reply:
Done